cxi/cxi_sampling_plugin.c

```diff
@@ -0,0 +1,344 @@
+      .fd      = fd,
+      .kind    = kind,
+  };
+  snprintf(fds[n_fds-1].ifname, sizeof fds[n_fds-1].ifname, "%s", ifname);
+
+  return 0;
+}
+
+static void open_all_fds(void) {
+  /* enumerate CXI devices once */
+  DIR *d = opendir(cxi_base);
+  if (!d) {
+    return;
+  }
+
+  struct dirent *de;
+  char path[PATH_MAX];
+
+  while ((de = readdir(d))) {
+    if (de->d_type != DT_LNK || de->d_name[0] == '.') {
+      continue;
+    }
+
+    const char *ifname = de->d_name;
+
+    for (size_t i = 0; i < n_counters; ++i) {
+      const char *cn = counter_names[i];
+
+      if (!strncmp(cn, "proc:", 5)) {
+        continue;                               /* later       */
+      }
+
+      if (!strncmp(cn, "rh:", 3)) {             /* RH counter  */
+        snprintf(path, sizeof(path),
+                 "%s/%s/%s",
+                 rh_base, ifname, cn + 3);
+        add_fd(path, ifname, cn, C_RH);
+      } else {                                  /* CXI counter */
+        snprintf(path, sizeof(path),
+                 "%s/%s/device/telemetry/%s",
+                 cxi_base, ifname, cn);
+        add_fd(path, ifname, cn, C_CXI);
+      }
+    }
+  }
+  closedir(d);
+
+  /* PROC counters - single per process */
+  for (size_t i = 0; i < n_counters; ++i) {
+    if (!strncmp(counter_names[i], "proc:", 5)) {
+      add_fd("/dev/null", "proc", counter_names[i], C_PROC); /* dummy fd */
+    }
+  }
+}
+
+static void close_all_fds(void) {
+  for (size_t i = 0; i < n_fds; ++i) {
+    if (fds[i].kind != C_PROC) {
+      close(fds[i].fd);
+    }
+  }
+  free(fds);
+  fds   = NULL;
+  n_fds = 0;
+}
+
+/* ------------------------------------------------------------------ */
+/*                3.  Fast sampling function                          */
+/* ------------------------------------------------------------------ */
+
+static int read_u64_pread(int fd, uint64_t *val) {
+  char buf[64];
+  ssize_t n = pread(fd, buf, sizeof(buf)-1, 0);
+  if (n <= 0) {
+    return -1;
+  }
+  buf[n] = '\0';
+
+  if (buf[0] == '@') {
+    return -1;         /* shouldn't happen */
+  }
+
+  /* CXI counters have “value@sec.nsec” - cut at '@' if present */
+  char *at = strchr(buf, '@');
+  if (at) {
+    *at = '\0';
+  }
+
+  errno = 0;
+  unsigned long long tmp = strtoull(buf, NULL, 0);
+  if (errno) {
+    return -1;
+  }
+  *val = (uint64_t)tmp;
+  return 0;
+}
+
+void thapi_sampling_cxi(void) {
+  struct rusage ru;
+  getrusage(RUSAGE_SELF, &ru);      /* PROC values once per loop */
+
+  for (size_t i = 0; i < n_fds; ++i) {
+    struct fd_entry *e = &fds[i];
+    uint64_t v = 0;
+
+    if (e->kind == C_PROC) {      /* chearper open (no sysfs) */
+      if (!strcmp(e->counter, "proc:ivcsw")) {
+          v = ru.ru_nivcsw;
+      } else if (!strcmp(e->counter, "proc:vcsw")) {
+          v = ru.ru_nvcsw;
+      } else {
+        continue;
+      }
+    } else if (read_u64_pread(e->fd, &v) < 0) {
+      continue;                 /* silent drop on error */
+    }
+
```

Comment:
why not finale `else ` continue? or `else {unreachable}`   I suppose we don't want to send a tracepoint with a balue of `0`

Reply:
Ahh, I get what you're saying now. Needed more coffee yesterday. I'll add the unreachable path and a few comments.

Reply:
Resolved in c653c22. Let me know if this is good enough for you. 😄

Reply:
Actually resolved in 7f709c5. Should be fallthrough. Logic 101 failure...

Reply:
Oh no I see... `read_u64_pread(e->fd, &v)` this set `v`...

Reply:
```
        if (e->kind == C_PROC) {      /* cheaper open (no sysfs) */
            if (!strcmp(e->counter, "proc:ivcsw")) {
                v = ru.ru_nivcsw;
            } else if (!strcmp(e->counter, "proc:vcsw")) {
                v = ru.ru_nvcsw;
            } else {
                continue;             /* unhandled proc counter */
            }
        }
        else {
            if (read_u64_pread(e->fd, &v) < 0) {
               continue;                 /* I/O error or parse error */
           } /* Read successfull, v is set */
        }
```
Ok this I think make more sense. Se `else if` was too smart  for my little brain. 
And if we wrote it like those handle alse the base